### ISSUE TITLE
Form-Validator: messages in small tags  is immediately overwritten

### ISSUE THREAD

**Issue by Zobakka** [bradtraversy/vanillawebprojects#28]
The `checkLength ` and `checkEmail  `functions inside "Event listeners" follow immediately after the `checkRequired` function.

If the input field is empty, the "field is required" message is immediately overwritten with a message like "user name must be at least 3 characters".

My solution is to return true or false for the showSuccess and showError functions and, depending on the result of the `if(checkRequired(username, email, password, password2))` to execute `checkLength ` and `checkEmail`.

**Comment by OmdaMukhtar** [comment]
Hi, @Zobakka I want to be PR with your repo, so I notice your issue is easy to implement, and depending on your description I did this solution could you have took look and let me know if it's correct, I tested on the browser and it's working fine!

```js
// Check required fields
function checkRequired(inputArr) {
  let isRequired = false;
  inputArr.forEach(function(input) {
    if (input.value.trim() === '') {
      showError(input, `${getFieldName(input)} is required`);
      isRequired = true;
    } else {
      showSuccess(input);
    }
  });

  return isRequired;
}
```

```js
// Event listeners
form.addEventListener('submit', function(e) {
  e.preventDefault();

  if(! checkRequired([username, email, password, password2])){
    checkLength(username, 3, 15);
    checkLength(password, 6, 25);
    checkEmail(email);
    checkPasswordsMatch(password, password2);
  }

});
```

**Comment by bradtraversy** [comment]
Please make a pull request and I will take a look at it

**Comment by OmdaMukhtar** [comment]
Hi @bradtraversy I have done https://github.com/bradtraversy/vanillawebprojects/pull/41

**Comment by Teledevhightech** [comment]
Hi @OmdaMukhtar  if you use isRequired checkRequired() won't return the correct check.
Because checkRequired takes an array and will check all items.
finally it will be confused with others functions check.

**Comment by Teledevhightech** [comment]
@OmdaMukhtar take look at:
const form = document.getElementById("form");
const username = document.getElementById("username");
const email = document.getElementById("email");
const password = document.getElementById("password");
const password2 = document.getElementById("password2");
//Fonction showSuccess
function showSuccess(input) {
  const formControl = input.parentElement;
  //console.log(formControl);
  formControl.className = "form-control success";
}
//Fonction showError
function showError(input, message) {
  const formControl = input.parentElement;
  //console.log(formControl);
  formControl.className = "form-control error";
  const small = formControl.querySelector("small");
  small.innerText = message;
}
// Fonction check input
function checkRequired(input) {
  return showError(input, `${getFieldName(input)} is required`);
}
//fonction qui verifie si l'email est valide
function checkEmail(input) {
  const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  if (input.value.trim() !== "") {
    if (re.test(input.value.trim())) {
      showSuccess(input);
    } else {
      showError(input, `${getFieldName(input)} is not valid`);
    }
  } else {
    checkRequired(input);
  }
}
//fonction qui recupere le du champ
function getFieldName(input) {
  return input.id.charAt(0).toUpperCase() + input.id.slice(1);
}
//fonction qui obtient la taille des caracteres saisis
function checkInputLength(input, min, max) {
  if (input.value.trim() !== "") {
    if (input.value.length < min) {
      showError(
        input,
        `${getFieldName(input)} must be at least ${min} characters`
      );
    } else if (input.value.length > max) {
      showError(input, `${getFieldName(input)} must be less than ${max}`);
    } else {
      showSuccess(input);
    }
  } else {
    checkRequired(input);
  }
}
function checkPasswordsMacth(input1, input2) {
  if (input2.value.trim() !== "") {
    if (input1.value === input2.value) {
      showSuccess(input2);
    } else {
      showError(input2, "Passwords not match");
    }
  } else {
    checkRequired(input2);
  }
}
//Fonction qui recupere le click
form.addEventListener("submit", function (e) {
  e.preventDefault();
  checkInputLength(username, 6, 15);
  checkInputLength(password, 6, 25);
  checkEmail(email);
  checkPasswordsMacth(password, password2);
});

**Comment by Teledevhightech** [comment]
const form = document.getElementById("form");
const username = document.getElementById("username");
const email = document.getElementById("email");
const password = document.getElementById("password");
const password2 = document.getElementById("password2");
//Fonction showSuccess
function showSuccess(input) {
  const formControl = input.parentElement;
  //console.log(formControl);
  formControl.className = "form-control success";
}
//Fonction showError
function showError(input, message) {
  const formControl = input.parentElement;
  //console.log(formControl);
  formControl.className = "form-control error";
  const small = formControl.querySelector("small");
  small.innerText = message;
}
// Fonction check input
function checkRequired(input) {
  return showError(input, `${getFieldName(input)} is required`);
}
//fonction qui verifie si l'email est valide
function checkEmail(input) {
  const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  if (input.value.trim() !== "") {
    if (re.test(input.value.trim())) {
      showSuccess(input);
    } else {
      showError(input, `${getFieldName(input)} is not valid`);
    }
  } else {
    checkRequired(input);
  }
}
//fonction qui recupere le du champ
function getFieldName(input) {
  return input.id.charAt(0).toUpperCase() + input.id.slice(1);
}
//fonction qui obtient la taille des caracteres saisis
function checkInputLength(input, min, max) {
  if (input.value.trim() !== "") {
    if (input.value.length < min) {
      showError(
        input,
        `${getFieldName(input)} must be at least ${min} characters`
      );
    } else if (input.value.length > max) {
      showError(input, `${getFieldName(input)} must be less than ${max}`);
    } else {
      showSuccess(input);
    }
  } else {
    checkRequired(input);
  }
}
function checkPasswordsMacth(input1, input2) {
  if (input2.value.trim() !== "") {
    if (input1.value === input2.value) {
      showSuccess(input2);
    } else {
      showError(input2, "Passwords not match");
    }
  } else {
    checkRequired(input2);
  }
}
//Fonction qui recupere le click
form.addEventListener("submit", function (e) {
  e.preventDefault();
  checkInputLength(username, 6, 15);
  checkInputLength(password, 6, 25);
  checkEmail(email);
  checkPasswordsMacth(password, password2);
});

**Comment by abraarahmed** [comment]
@OmdaMukhtar There was an issue with your fix #41, the validation check for username and password length stopped working. Corrected fix #56. @bradtraversy Please have a look.

**Comment by Zobakka** [comment]
Hi everyone! Haven't seen all your messages because was deep inside CV19.

Unfortunately all your solutions don't work.

1. Current version (after commit 851efa8) doesn't check fields length and pass email validation. 

It happens besause of forEach loop in the checkRequired function. let's experiment a little.

Fill the first input with 1. Press Sumbit button. Foreach loop gets the first input value, leaves isRequired = false, runs showSuccess(username) goes through all fields, and returns true, because last input is empty and sets isRequired = false.

As result if(!checkRequired([username, email, password, password2])) statement will be false and checklength functions will never run. It happens only if the last field will be filled.

2. The  abraarahmed fix #56 works in another way.

If press Submit button with empty fields, or fill all the fields in the right way is everything OK - "all field are required" or checks passed.

But fill in just one any field -  forEach loop in the checkRequired function goes through all inputs, sets once isRequired = false, runs showError for the empty inputs and sets Error message into "field is requered", returns false.

if(!checkRequired([username, email, password, password2]) statement becames true and Error messages immediately overrides by checkLength functions.

Please look at current version (https://codepen.io/konstantin-dotsenko/pen/ZEWQQzr), abraarahmed (https://codepen.io/konstantin-dotsenko/pen/gOrPPzm)  and my (https://codepen.io/konstantin-dotsenko/pen/JjXGYqe)

**Comment by abraarahmed** [comment]
@Zobakka You are right. In my fix #56, the individual field validation messages overrode all the required (non empty) validation messages, even if one field was filled. Your suggestion fixes this, @bradtraversy please check.

_P.S.: I have closed my pull requests._
